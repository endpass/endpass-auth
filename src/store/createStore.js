// @ts-check
import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

/**
 * @param {object?} [options] passed options for store
 * @return {import('vuex').Store<{}>} store Vuex store instance
 */
export default options => {
  return new Vuex.Store({
    strict: !ENV.VUE_APP_IS_PRODUCTION,
    ...options,
  });
}
