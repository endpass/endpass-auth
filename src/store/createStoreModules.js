import GasPriceModule from '@/store/modules/GasPriceModule';
import SharedModule from '@/store/modules/SharedModule';
import ChannelStore from '@/store/modules/ChannelModule';
import BalanceModule from '@/store/modules/BalanceModule';
import AccountsModule from '@/store/modules/AccountsModule';
import AuthModule from '@/store/modules/AuthModule';
import CoreModule from '@/store/modules/CoreModule';
import RequestsModule from '@/store/modules/RequestsModule';
import moduleFactory from '@/store/moduleFactory';

export default store => {
  const createModule = moduleFactory(store);

  const gasPriceStore = createModule(GasPriceModule, 'gasPrice');

  const sharedStore = createModule(SharedModule, 'shared');

  const channelStore = createModule(ChannelStore, 'channel');

  const balanceStore = createModule(BalanceModule, 'balance');

  const accountsStore = createModule(AccountsModule, 'accounts', {
    sharedStore,
    balanceStore,
  });

  const authStore = createModule(AuthModule, 'auth', {
    sharedStore,
  });

  const coreStore = createModule(CoreModule, 'core', {
    authStore,
    sharedStore,
  });

  const requestStore = createModule(RequestsModule, 'requests', {
    accountsStore,
    sharedStore,
  });

  return {
    channelStore,
    gasPriceStore,
    sharedStore,
    accountsStore,
    balanceStore,
    authStore,
    coreStore,
    requestStore,
  };
};
