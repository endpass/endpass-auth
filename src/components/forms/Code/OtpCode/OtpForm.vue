<template>
  <form @submit.prevent="onSubmit">
    <v-title>
      {{ $t('components.otp.title') }}
    </v-title>
    <v-description>
      {{ $t('components.otp.description') }}
    </v-description>

    <form-item>
      <v-code-input
        v-model="code"
        v-validate="'required|digits:6'"
        class="v-text-center"
        data-vv-as="code"
        data-vv-name="code"
        :error="errors.first('code') || error"
        name="code"
        :placeholder="$t('components.otp.enterReceivedCode')"
        data-test="code-input"
      />
    </form-item>
    <form-item class="v-mb-24">
      <v-button
        :disabled="!isFormValid || isLoading"
        :is-loading="isLoading"
        type="submit"
        data-test="submit-button"
      >
        {{ $t('global.confirm') }}
      </v-button>
    </form-item>
    <form-row class="v-fs-14 v-text-center">
      <v-link
        :disabled="isLoading"
        href="#"
        data-test="recovery-link"
        @click.prevent="onRecover"
      >
        {{ $t('components.otp.noCode') }}
      </v-link>
    </form-row>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import VCodeInput from '@endpass/ui/kit/VCodeInput';
import VLink from '@endpass/ui/kit/VLink';
import { authStore } from '@/store';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';
import VDescription from '@/components/common/VDescription';

export default {
  name: 'OtpForm',

  authStore,

  props: {
    email: {
      type: String,
      required: true,
    },

    password: {
      type: String,
      required: true,
    },

    isSignUp: {
      type: Boolean,
      required: true,
    },

    submitHandler: {
      type: Function,
      required: true,
    },
  },

  data: () => ({
    code: '',
    error: null,
    isLoading: false,
  }),

  methods: {
    async onSubmit() {
      if (this.isLoading) return;
      try {
        const { code, email, password, isSignUp } = this;
        this.isLoading = true;
        this.error = null;

        await this.submitHandler({
          isSignUp,
          email,
          password,
          code,
        });

        this.$emit('submit', code);
      } catch (err) {
        this.error = this.$i18n.t('components.otpBlock.authFailed');
      } finally {
        this.isLoading = false;
      }
    },

    onRecover() {
      if (this.isLoading) return;

      this.$emit('recover');
    },
  },

  mixins: [formMixin],

  components: {
    VTitle,
    VDescription,
    VLink,
    VButton,
    VCodeInput,
    FormItem,
    FormRow,
  },
};
</script>

<style lang="postcss"></style>
