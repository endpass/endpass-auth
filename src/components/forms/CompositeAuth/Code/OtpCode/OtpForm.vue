<template>
  <form
    class="form-otp"
    @submit.prevent="onSubmit"
  >
    <v-title>
      <span v-html="$t('components.otp.title')" />
    </v-title>
    <v-description>
      <span v-html="$t('components.otp.description')" />
    </v-description>

    <form-item>
      <v-input
        v-model="code"
        v-validate="'required|digits:6'"
        data-vv-as="code"
        data-vv-name="code"
        :error="errors.first('code') || error"
        name="code"
        :placeholder="$t('components.otp.enterReceivedCode')"
        data-test="code-input"
      />
    </form-item>
    <form-item class="v-mb-24">
      <v-button
        :disabled="!isFormValid || isLoading"
        type="submit"
        data-test="submit-button"
      >
        {{ primaryButtonLabel }}
      </v-button>
    </form-item>
    <form-row
      class="v-fs-14"
      centered
    >
      <v-link
        :disabled="isLoading"
        href="#"
        data-test="recovery-link"
        @click.prevent="emitRecoverEvent"
      >
        {{ $t('components.otp.noCode') }}
      </v-link>
    </form-row>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import VInput from '@endpass/ui/kit/VInput';
import VLink from '@endpass/ui/kit/VLink';
import { authStore } from '@/store';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';
import VDescription from '@/components/common/VDescription';

export default {
  name: 'OtpForm',

  authStore,

  props: {
    email: {
      type: String,
      required: true,
    },

    password: {
      type: String,
      required: true,
    },
  },

  data: () => ({
    code: '',
    error: '',
    isLoading: false,
  }),

  computed: {
    primaryButtonLabel() {
      return !this.isLoading
        ? this.$i18n.t('global.confirm')
        : this.$i18n.t('global.loading');
    },
  },

  methods: {
    async onSubmit() {
      try {
        const { code, email, password } = this;
        this.isLoading = true;
        this.error = null;
        await this.$options.authStore.authByCode({
          email,
          password,
          code,
        });

        this.$emit('submit', code);
      } catch (err) {
        // TODO: add handling
        this.error = err;
      } finally {
        this.isLoading = false;
      }
    },

    emitRecoverEvent() {
      if (this.isLoading) return;

      this.$emit('recover');
    },
  },

  mixins: [formMixin],

  components: {
    VTitle,
    VDescription,
    VLink,
    VButton,
    VInput,
    FormItem,
    FormRow,
  },
};
</script>

<style lang="postcss"></style>
