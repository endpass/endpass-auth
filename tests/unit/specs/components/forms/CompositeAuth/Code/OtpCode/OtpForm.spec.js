import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import OtpForm from '@/components/forms/CompositeAuth/Code/OtpCode/OtpForm';
import setupI18n from '@/locales/i18nSetup';
import createStore from '@/store/createStore';
import createStoreModules from '@/store/createStoreModules';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VeeValidate);
const i18n = setupI18n(localVue);

describe('Otp', () => {
  let wrapper;
  let coreStore;

  beforeEach(() => {
    jest.clearAllMocks();

    const store = createStore();
    const { coreStore: coreStoreModule } = createStoreModules(store);

    coreStore = coreStoreModule;

    wrapper = shallowMount(OtpForm, {
      coreStore,
      localVue,
      provide: {
        theme: 'default',
      },
      i18n,
      sync: false,
    });
  });

  describe('render', () => {
    it('should correctly render Otp component', () => {
      expect(wrapper.name()).toBe('OtpForm');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should correctly disable recovery link', async () => {
      expect.assertions(2);

      const recoveryLink = wrapper.find('[data-test=recovery-link]');

      coreStore.changeLoadingStatus(false);
      await wrapper.vm.$nextTick();

      expect(recoveryLink.attributes().disabled).toBeUndefined();

      coreStore.changeLoadingStatus(true);
      await wrapper.vm.$nextTick();

      expect(recoveryLink.attributes().disabled).toBeTruthy();
    });
  });

  describe('behavior', () => {
    it('should not emit recover event', () => {
      coreStore.changeLoadingStatus(true);

      wrapper.find('[data-test=recovery-link]').vm.$emit('click', {
        preventDefault: () => {},
      });

      expect(wrapper.emitted().recover).toBe(undefined);
    });

    it('should emit recover event', () => {
      coreStore.changeLoadingStatus(false);

      wrapper.find('[data-test=recovery-link]').vm.$emit('click', {
        preventDefault: () => {},
      });

      expect(wrapper.emitted().recover).toEqual([[]]);
    });
  });
});
