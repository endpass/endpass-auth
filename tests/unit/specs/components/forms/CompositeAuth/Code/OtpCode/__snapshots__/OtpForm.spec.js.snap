// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Otp render should correctly render Otp component 1`] = `
<form class="form-otp">
  <v-title-stub html="Verify OTP code" data-test="form-title"></v-title-stub>
  <v-description-stub html="Enter the 6&amp;nbsp;digit code generated by&amp;nbsp;your authenticator app."></v-description-stub>
  <form-item-stub>
    <v-input-stub value="" data-vv-as="code" data-vv-name="code" name="code" placeholder="Enter received code..." data-test="code-input"></v-input-stub>
  </form-item-stub>
  <form-item-stub islast="true">
    <v-button-stub skin="primary" size="normal" type="submit" data-test="submit-button" disabled="true">
      Confirm
    </v-button-stub>
  </form-item-stub>
  <form-row-stub centered="" class="v-text-size-14">
    <v-link-stub href="#" skin="default" target="_blank" data-test="recovery-link">
      Don't know code?
    </v-link-stub>
  </form-row-stub>
</form>
`;
